### PR TITLE
[WIP] group by without aggregations with support for sql and pandas style

### PR DESCRIPTION
**Pull request by williamBlazing**
This PR supersedes PR #430  (https://github.com/rapidsai/cudf/pull/430) . I created a new PR mostly due to local branch git issues. I will take all recommendations from that PR into account. For now I would like some feedback on the implementation, in particular about the new gdf_context fields I am now using, how the gdf_group_by_wo_aggregations is organized and its usage and addition of a gdf_group_start_indices function.

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -50,7 +50,7 @@ set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Werror cross-execution-space-call -Xc
 set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Xcompiler -D_GLIBCXX_USE_CXX11_ABI=0")
 
 # set default build type
-set(CMAKE_BUILD_TYPE "Release")
+set(CMAKE_BUILD_TYPE "Debug")
```
Comment (eyalroz) [inline cpp/CMakeLists.txt]: This doesn't seem to belong in this patch.
Also, people downloading and building cudf expect a release build, not a debug build

File: cpp/include/cudf/functions.h [hunk 2]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
                              gdf_column* out_col_agg,      //aggregation result
                              gdf_context* ctxt);            //struct with additional info: bool is_sorted, flag_sort_or_hash, bool flag_count_distinct
 
+/* --------------------------------------------------------------------------*/
+  /**
+   * @brief Given a set of columns, of which a subset of these are defined to be the group by columns,
```
Comment (eyalroz) [inline cpp/include/cudf/functions.h]: "of which a subset of these" <- mis-phrased.

File: cpp/include/cudf/functions.h [hunk 3]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
+   * all columns are sorted by the group by columns and returned, along with a column containing
+   * a list of the start indices of each group
+   *
+   * @Param[in] The number of columns in the dataset
```
Comment (eyalroz) [inline cpp/include/cudf/functions.h]: Why doesn't this have the parameter name for each parameter?

File: cpp/include/cudf/functions.h [hunk 4]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
+   * @Param[in] The input columns in the dataset
+   * @Param[in] The number of columns to be grouping by
+   * @Param[in] The column indices of the input dataset that will be grouped by
+   * @Param[out] The dataset sorted by the group by columns (needs to be pre-allocated)
+   * @Param[out] A column containing the starting indices of each group. Indices based off of new sort order. (needs to be pre-allocated)
```
Comment (eyalroz) [inline cpp/include/cudf/functions.h]: Somewhat confusing:
* We already have at least two kinds of indices: An index into a data and an index into the sequence of columns. Is this a new kind?
* "Indices based off of new sort order" - This is not a sentence. Even if it were to say "indices are based off of the new sort order" - this is still not clear. Offsets are "based off of" things. "New sort order" - it's not clear to me what order that is exactly.
* "Needs to be pre-allocated" - How much space is necessary exactly?

File: cpp/include/cudf/functions.h [hunk 5]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
+   * @Param[in] The context used to control how nulls are treated in a sort and in group by
+   *   context->flag_nulls_sort_behavior< 0 = Nulls are are treated as largest, 
+   *   1 = Nulls are treated as smallest, 2 = Special multi-sort case any row with null is largest>
+   *   context-> flag_groupby_include_nulls <0 = Nulls are ignored in group by keys (Pandas style), 
+                        1 = Nulls are treated as values in group by keys where NULL == NULL (SQL style)>
+   *
+   * @Returns gdf_error with error code on failure, otherwise GDF_SUCESS
+   */
+  /* ----------------------------------------------------------------------------*/
+gdf_error gdf_group_by_wo_aggregations(int num_data_cols,
```
Comment (eyalroz) [inline cpp/include/cudf/functions.h]: So, I've looked at this function's Doxygen comment, and - this does not perform GROUP BY. It is merely an out-of-place sort.

Also:
1. Shorthands are confusing for many people who can't guess what you meant. Use "without" rather than "wo".
2. The type of `num_data_cols` should not be a plain `int`, we should have a type definition for a column index / number-of-columns.

File: cpp/include/cudf/functions.h [hunk 6]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
                              gdf_context* ctxt);            //struct with additional info: bool is_sorted, flag_sort_or_hash, bool flag_count_distinct
 
+/* --------------------------------------------------------------------------*/
+  /**
+   * @brief Given a set of columns, of which a subset of these are defined to be the group by columns,
+   * all columns are sorted by the group by columns and returned, along with a column containing
+   * a list of the start indices of each group
+   *
+   * @Param[in] The number of columns in the dataset
+   * @Param[in] The input columns in the dataset
```
Comment (eyalroz) [inline cpp/include/cudf/functions.h]: Need to mention these must all have the same number of elements (the same "size").

File: cpp/include/cudf/functions.h [hunk 7]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
+   *   context-> flag_groupby_include_nulls <0 = Nulls are ignored in group by keys (Pandas style), 
+                        1 = Nulls are treated as values in group by keys where NULL == NULL (SQL style)>
+   *
+   * @Returns gdf_error with error code on failure, otherwise GDF_SUCESS
+   */
+  /* ----------------------------------------------------------------------------*/
+gdf_error gdf_group_by_wo_aggregations(int num_data_cols,
+                           	   	   	   gdf_column** data_cols_in,
+									   int num_groupby_cols,
+									   int * groupby_col_indices,
```
Comment (eyalroz) [inline cpp/include/cudf/functions.h]: I'd try to choose a clearer name. Also, why are the indices `int`s? If I followed the doxygen correctly, shouldn't these be `gdf_size`'s?

File: cpp/include/cudf/functions.h [hunk 8]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
+   * a list of the start indices of each group
+   *
+   * @Param[in] The number of columns in the dataset
+   * @Param[in] The input columns in the dataset
+   * @Param[in] The number of columns to be grouping by
+   * @Param[in] The column indices of the input dataset that will be grouped by
+   * @Param[out] The dataset sorted by the group by columns (needs to be pre-allocated)
+   * @Param[out] A column containing the starting indices of each group. Indices based off of new sort order. (needs to be pre-allocated)
+   * @Param[in] The context used to control how nulls are treated in a sort and in group by
+   *   context->flag_nulls_sort_behavior< 0 = Nulls are are treated as largest,
```
Comment (eyalroz) [inline cpp/include/cudf/functions.h]: The description of NULL behavior during a sort context setting should not appear here: It should appear in the Doxygen for `gdf_context`; here, we should merely indicate which aspects of the context we're using.

Also - no magic numbers in general, and in comments particularly.

File: cpp/include/cudf/functions.h [hunk 9]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
+  /**
+   * @brief Given a set of columns, of which a subset of these are defined to be the group by columns,
+   * all columns are sorted by the group by columns and returned, along with a column containing
+   * a list of the start indices of each group
+   *
+   * @Param[in] The number of columns in the dataset
+   * @Param[in] The input columns in the dataset
+   * @Param[in] The number of columns to be grouping by
+   * @Param[in] The column indices of the input dataset that will be grouped by
+   * @Param[out] The dataset sorted by the group by columns (needs to be pre-allocated)
```
Comment (eyalroz) [inline cpp/include/cudf/functions.h]: Clarify what exactly needs to be pre-allocated, and how many elements.

File: cpp/include/cudf/types.h [hunk 10]
```diff
@@ -165,6 +165,10 @@ typedef struct gdf_context_{
   int flag_distinct;            /**< for COUNT: DISTINCT = 1, else = 0 */
   int flag_sort_result;         /**< When method is GDF_HASH, 0 = result is not sorted, 1 = result is sorted */
   int flag_sort_inplace;        /**< 0 = No sort in place allowed, 1 = else */
+  int flag_groupby_include_nulls; /**< 0 = Nulls are ignored in group by keys (Pandas style),
```
Comment (eyalroz) [inline cpp/include/cudf/types.h]: 1. Boolean flags should be `bool`s. 
2. No magic numbers in comments. If it were a bool flag, with a proper name, then even without a comment it would be relatively clear what it means.
3. What does it mean that "NULLs are ignored by group by keys" - does that mean the tuples with NULLs are dropped? Not clear.
4. Instead of just saying what the pandas-style is, there should be some sort of mechanism for saying "handle nulls pandas-style, e.g. some kind of struct/class which, when applied to a gdf_context, sets this flag to have its pandas-style value.

File: cpp/include/cudf/types.h [hunk 11]
```diff
@@ -165,6 +165,10 @@ typedef struct gdf_context_{
   int flag_distinct;            /**< for COUNT: DISTINCT = 1, else = 0 */
   int flag_sort_result;         /**< When method is GDF_HASH, 0 = result is not sorted, 1 = result is sorted */
   int flag_sort_inplace;        /**< 0 = No sort in place allowed, 1 = else */
+  int flag_groupby_include_nulls; /**< 0 = Nulls are ignored in group by keys (Pandas style), 
+                                      1 = Nulls are treated as values in group by keys where NULL == NULL (SQL style)*/ 
+  int flag_nulls_sort_behavior; /**< 0 = Nulls are are treated as largest, 1 = Nulls are treated as smallest,
```
Comment (eyalroz) [inline cpp/include/cudf/types.h]: Multi-valued options, which are not meaningful as actual numbers, should use enum classes. If they're supposed to be C-compatible - then plain enums.

Also, same comment regarding the magic numbers. Yes, I know other fields in `gdf_cotext` may already have this affliction...

File: cpp/src/bitmask/bitmask_ops.cu [hunk 12]
```diff
@@ -4,8 +4,8 @@
 #include "cudf/functions.h"
 #include "rmm/thrust_rmm_allocator.h"
 
-#include <cuda_runtime.h>
 #include <vector>
+#include <cuda_runtime.h>
```
Comment (eyalroz) [inline cpp/src/bitmask/bitmask_ops.cu]: Actually, there is a "one true include order" which we should follow throughout cudf. But never mind, fine for now.

File: cpp/src/bitmask/valid_ops.cu [hunk 13]
```diff
@@ -186,37 +189,59 @@ gdf_error gdf_count_nonzero_mask(gdf_valid_type const * masks, int num_rows, int
...
   {
-    // TODO: Probably shouldn't create/destroy the stream every time
-    cudaStream_t count_stream;
-    CUDA_TRY(cudaStreamCreate(&count_stream));
-    int * d_count{nullptr};
+    int* d_count{nullptr};
 
     // Cast validity buffer to 4 byte type
-    valid32_t const * masks32{reinterpret_cast<valid32_t const *>(masks)};
+    valid32_t const* masks32{reinterpret_cast<const valid32_t*>(masks)};
```
Comment (eyalroz) [inline cpp/src/bitmask/valid_ops.cu]: Please check whether you can make do with a static cast.

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (jrhemstad) [reply cpp/CMakeLists.txt]: Agreed.

File: cpp/include/cudf/functions.h (hunk 2 above)
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]: I find it grammatically easier to refer to the columns you're "grouping by" as the "key" columns rather than "group by" columns.

File: cpp/include/cudf/functions.h [hunk 14]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
 
+/* --------------------------------------------------------------------------*/
+  /**
+   * @brief Given a set of columns, of which a subset of these are defined to be the group by columns,
+   * all columns are sorted by the group by columns and returned, along with a column containing
+   * a list of the start indices of each group
+   *
+   * @Param[in] The number of columns in the dataset
+   * @Param[in] The input columns in the dataset
+   * @Param[in] The number of columns to be grouping by
```
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]: e.g., "The number of key columns" is easier than "The number of columns to be grouping by".

File: cpp/include/cudf/functions.h (hunk 5 above)
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]: I agree this API name is messy. I think it should simply be `gdf_groupby` and the existing `gdf_groupby` should be renamed `gdf_groupby_aggregation`. This better matches the semantics of Pandas where a standalone `groupby` doesn't perform an aggregation by default, it just returns the groups (same as your function here).

File: cpp/src/bitmask/valid_ops.cu [hunk 15]
```diff
@@ -167,10 +167,13 @@ void count_valid_bits(valid32_t const * const masks32,
  * @Returns  GDF_SUCCESS upon successful completion 
  *
  * ----------------------------------------------------------------------------*/
-gdf_error gdf_count_nonzero_mask(gdf_valid_type const * masks, int num_rows, int * count)
+gdf_error count_nonzero_mask(gdf_valid_type const * masks,
+                                 int num_rows,
+                                 int& count,
```
Comment (jrhemstad) [inline cpp/src/bitmask/valid_ops.cu]: Output parameters passed by reference are potentially confusing at the call site. Much clearer to pass output parameters by pointer.

File: cpp/include/cudf/functions.h [hunk 16]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
+   * @Param[in] The column indices of the input dataset that will be grouped by
+   * @Param[out] A column containing the starting indices of each group. Indices based off of new sort order. (needs to be pre-allocated)
+   * @Param[in] The context used to control how nulls are treated in a sort
+   *   context->flag_nulls_sort_behavior< 0 = Nulls are are treated as largest, 
+   *   1 = Nulls are treated as smallest, 2 = Special multi-sort case any row with null is largest>
+   *
+   * @Returns gdf_error with error code on failure, otherwise GDF_SUCESS
+   */
+  /* ----------------------------------------------------------------------------*/
+gdf_error gdf_group_start_indices(int num_data_cols,
```
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]: I'm thinking that there's a more intuitive and generic name for this function. It's very similar to `thrust::unique`, except it doesn't modify your input, it just gives you an offset to the start of each unique group.

Brainstorming...

`gdf_unique`
`gdf_unique_indices`
...

File: cpp/include/cudf/functions.h [hunk 17]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
+                           	   	   	   gdf_column** data_cols_in,
+									   int num_groupby_cols,
+									   int * groupby_col_indices,
+									   gdf_column** data_cols_out,
+									   gdf_column* group_start_indices,
+                     gdf_context* ctxt);           
+
+/* --------------------------------------------------------------------------*/
+  /**
+   * @brief Given a set of columns, of which a subset of these are defined to be the group by columns,
```
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]: See other comments about using "key" columns instead of "group by" columns.

File: cpp/include/cudf/functions.h [hunk 18]
```diff
@@ -894,6 +894,60 @@ gdf_error gdf_group_by_count(int ncols,                    // # columns
...
+   * all input data is assumed to already be sorted by these group by columns. 
+   * This function calculates a list of the start indices of each group
+   *
+   * @Param[in] The number of columns in the dataset (assumed to already be sorted)
```
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]: Documentation is missing the parameter names

File: cpp/include/cudf/types.h (hunk 11 above)
Comment (jrhemstad) [reply cpp/include/cudf/types.h]: Agreed, a `typedef enum` would be appropriate here.

File: cpp/src/dataframe/cudf_table.cuh [hunk 19]
```diff
@@ -370,6 +371,10 @@ public:
     return row_valid;
   }
 
+  gdf_error get_num_valid_rows(int & num_valid_rows) {
```
Comment (jrhemstad) [inline cpp/src/dataframe/cudf_table.cuh]: Output parameter passed by reference.

File: cpp/src/groupby/new_groupby.cu [hunk 20]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
 
   return gdf_error_code;
 }
+
+
+// thrust::device_vector set to use rmmAlloc and rmmFree.
+template <typename T>
+using Vector = thrust::device_vector<T, rmm_allocator<T>>;
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Use `rmm::device_vector` instead.

File: cpp/src/groupby/new_groupby.cu [hunk 21]
```diff
@@ -6,6 +6,11 @@
 #include "aggregation_operations.hpp"
 #include "groupby/hash_groupby.cuh"
 
+#include <thrust/fill.h>
+
+#include <../tests/utilities/cudf_test_utils.cuh>
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Why is a unit test header being included here?

File: cpp/src/groupby/new_groupby.cu [hunk 22]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
...
+  GDF_REQUIRE((nullptr != data_cols_in), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((nullptr != data_cols_in[0]), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((nullptr != data_cols_out), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((nullptr != data_cols_out[0]), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((num_data_cols > 0), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((num_groupby_cols > 0), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((nullptr != groupby_col_indices), GDF_DATASET_EMPTY);
+
+
+  int32_t nrows = data_cols_in[0]->size;
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: There should be a check if the number of rows in the input is 0 to just return GDF_SUCCESS.

File: cpp/src/groupby/new_groupby.cu [hunk 23]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
...
+  }
+
+  rmm::device_vector<int32_t> sorted_indices(nrows);
+  gdf_column sorted_indices_col;
+  gdf_error status = gdf_column_view(&sorted_indices_col, (void*)(sorted_indices.data().get()), 
+                            nullptr, nrows, GDF_INT32);
+  if (status != GDF_SUCCESS)
+    return status;
+
+  if (ctxt->flag_groupby_include_nulls == 1 || !group_by_keys_contain_nulls){  // SQL style
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Consider wrapping the contents of the `if` in a function (in anonymous namespace) for the "SQL Style" (perhaps there's a better name than "SQL style").

File: cpp/src/groupby/new_groupby.cu [hunk 24]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
...
+
+    status = table_in->gather<int32_t>(sorted_indices, *table_out.get());
+    if (status != GDF_SUCCESS)
+      return status;
+
+    status = gdf_group_start_indices(num_data_cols, data_cols_out, num_groupby_cols,
+                      groupby_col_indices, group_start_indices, ctxt);
+
+    return status;
+  } else {  // Pandas style
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Similar to above. Consider wrapping the contents of this `else` in a function that captures the "Pandas style".

Perhaps there's a way to distill both functions into a single function controlled by a flag.

File: cpp/src/groupby/new_groupby.cu [hunk 25]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
...
+
+    int flag_nulls_sort_behavior = ctxt->flag_nulls_sort_behavior;
+    ctxt->flag_nulls_sort_behavior = 2; // overide behaviour to filter out the nulls
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Instead of overwriting the `context` that's passed in, just create a new `gdf_context` and copy the input `gdf_context` into your temporary.

But I guess since it's a C struct you don't get a default copy constructor...

File: cpp/src/groupby/new_groupby.cu [hunk 26]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
...
+                     gdf_column** data_cols_in,
+									   int num_groupby_cols,
+									   int * groupby_col_indices,
+									   gdf_column* group_start_indices,
+									   gdf_context* ctxt)
+{
+
+  int32_t nrows = data_cols_in[0]->size;
+  // setup for reduce by key  
+  bool have_nulls = false;
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Cleaner by using `std::any`
```suggestion
   bool const have_nulls{ std::any_of(data_cols_in, data_cols_in + num_groupby_cols, [](gdf_column * col){ return col->null_count > 0; }) };
```

File: cpp/src/groupby/new_groupby.cu [hunk 27]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
...
+    gdf_valid_type** d_valids_data = d_valids.data().get();
+
+    soa_col_info(data_cols_in, num_groupby_cols, d_col_data, d_valids_data, d_col_types);
+    
+    LesserRTTI<int32_t> comp(d_col_data, d_valids_data, d_col_types, nullptr, num_groupby_cols, nulls_are_smallest);
+    
+    auto counting_iter = thrust::make_counting_iterator<int32_t>(0);
+    
+    result_end = thrust::unique_copy(exec, counting_iter, counting_iter+nrows, 
+                              (int32_t*)group_start_indices->data,
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Avoid C-style casts.

File: cpp/src/groupby/new_groupby.cu [hunk 28]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
...
+  } else {
+
+    soa_col_info(*data_cols_in, num_groupby_cols, d_col_data, d_col_types);
+    
+    LesserRTTI<int32_t> comp(d_col_data, nullptr, d_col_types, nullptr, num_groupby_cols, nulls_are_smallest);
+
+    auto counting_iter = thrust::make_counting_iterator<int32_t>(0);
+    
+    result_end = thrust::unique_copy(exec, counting_iter, counting_iter+nrows, 
+                              (int32_t*)group_start_indices->data,
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Avoid C-style casts

File: cpp/src/groupby/new_groupby.cu [hunk 29]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
...
+                              [comp] __host__ __device__(int32_t key1, int32_t key2){
+                              return comp.equal(key1, key2);  
+                            });
+  }
+
+  size_t new_sz = thrust::distance((int32_t*)group_start_indices->data, result_end);
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Avoid C-style casts.

File: cpp/src/groupby/new_groupby.cu [hunk 30]
```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
...
+   *   1 = Nulls are treated as smallest, 2 = Special multi-sort case any row with null is largest>
+   *
+   * @Returns gdf_error with error code on failure, otherwise GDF_SUCESS
+   */
+  /* ----------------------------------------------------------------------------*/
+gdf_error gdf_group_start_indices(int num_data_cols,
+                     gdf_column** data_cols_in,
+									   int num_groupby_cols,
+									   int * groupby_col_indices,
+									   gdf_column* group_start_indices,
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: This code assumes `groupby_col_indices->dtype == GDF_INT32`. Make sure to add a `GDF_REQUIRE` for that.

File: cpp/include/cudf/types.h (hunk 10 above)
Comment (jrhemstad) [reply cpp/include/cudf/types.h]: You could define your own `bool` as a `typedef enum`.

Comment (eyalroz) [reply cpp/include/cudf/types.h]: @jrhemstad : C has had bools for nearly 20 years, actually. See: http://www.c-faq.com/bool/booltype.html

File: cpp/include/cudf/functions.h (hunk 2 above)
Comment (eyalroz) [reply cpp/include/cudf/functions.h]: @jrhemstad : Among database people, "key" is commonly used w.r.t. tables or joins, not grouping.

File: cpp/include/cudf/types.h (hunk 10 above)
Comment (jrhemstad) [reply cpp/include/cudf/types.h]: It's not native to the language. `types.h` cannot have any `include` statements because CFFI does not have a preprocessor.

File: cpp/include/cudf/functions.h (hunk 2 above)
Comment (jrhemstad) [reply cpp/include/cudf/functions.h]: But they are "key"s. Groupby is just a "reduce-by-key", where the "groupby" columns are the "key" columns. Pandas users I've talked to also refer to these as the key columns.

File: cpp/include/cudf/types.h (hunk 10 above)
Comment (eyalroz) [reply cpp/include/cudf/types.h]: @jrhemstad : The native-to-the-language version is `_Bool`. Of course, when including in C++ we would need to prepend `using _Bool = bool`;

but it's a bit baroque, so never mind.

File: cpp/include/cudf/functions.h (hunk 2 above)
Comment (harrism) [reply cpp/include/cudf/functions.h]: And we use this terminology ("keys") elsewhere in cuDF.

**Comment by harrism** [comment]
@williamBlazing There's a lot to do in this PR. I'm guessing it will not make the v0.5 code freeze next week. Please correct me if I'm wrong.

**Comment by aocsa** [comment]
@williamBlazing you can close this PR because I create a new one at
https://github.com/rapidsai/cudf/pull/982
